integration/ledger/snapshot_test.go

```diff
@@ -0,0 +1,1049 @@
+
+	By("verifying blockchain info on peer " + peer.ID())
+	sess, err := n.PeerUserSession(peer, "Admin", commands.ChannelInfo{
+		ChannelID: channelID,
+	})
+	Expect(err).NotTo(HaveOccurred())
+	Eventually(sess, n.EventuallyTimeout).Should(gexec.Exit(0))
+	channelInfoStr := strings.TrimPrefix(string(sess.Buffer().Contents()[:]), "Blockchain info:")
+	var bcInfo = common.BlockchainInfo{}
+	err = json.Unmarshal([]byte(channelInfoStr), &bcInfo)
+	Expect(err).NotTo(HaveOccurred())
+	Expect(bcInfo.Height).To(Equal(uint64(channelHeight)))
+	Expect(bcInfo.BootstrappingSnapshotInfo.LastBlockInSnapshot).To(Equal(uint64(lastBlockInSnapshot)))
+
+	By("waiting for all peers to have the same ledger height")
+	channelPeers := n.PeersWithChannel(channelID)
+	maxLedgerHeight := nwo.GetMaxLedgerHeight(n, channelID, channelPeers...)
+	nwo.WaitUntilEqualLedgerHeight(n, channelID, maxLedgerHeight, channelPeers...)
+}
+
+// verify the following functions for the peer after joining a channel by snapshot
+// - invoke/query chaincode
+// - upgrade chaincode
+// - channel config update
+// - join new channel
+// - qscc
+// - query history
+func verifyPeerAfterJoinBySnapshot(helper *marblesTestHelper, peer *nwo.Peer, channelID string, chaincode *nwo.Chaincode, newCCSequence string,
+	lastBlockInSnapshot int, oldTxid string, testKey string, availableHistoryResult []*marbleHistoryResult) []*marbleHistoryResult {
+
+	n := helper.Network
+	orderer := helper.orderer
+	peerID := peer.ID()
+
+	helper.invokeMarblesChaincode(chaincode.Name, peer, "transferMarble", testKey, "newowner_"+newCCSequence)
+
+	if newCCSequence != "" && newCCSequence != chaincode.Sequence {
+		chaincode.Version = newCCSequence + ".0"
+		chaincode.Sequence = newCCSequence
+		By("upgrading chaincode to version " + chaincode.Version + " on peer " + peerID)
+		nwo.DeployChaincode(n, channelID, orderer, *chaincode)
+	}
+
+	By("updating policy in channel config on peer " + peerID)
+	var org *nwo.Organization
+	for _, o := range n.Organizations {
+		if o.Name == peer.Organization {
+			org = o
+			break
+		}
+	}
+	currentConfig := nwo.GetConfig(n, peer, orderer, channelID)
+	updatedConfig := proto.Clone(currentConfig).(*common.Config)
+	updatedConfig.ChannelGroup.Groups["Application"].Groups[org.Name].Policies["Writers"].Policy.Value = protoutil.MarshalOrPanic(policydsl.SignedByMspMember(org.MSPID))
+	nwo.UpdateConfig(n, orderer, channelID, currentConfig, updatedConfig, true, peer)
+
+	By("joining a new channel testchannelID3 on peer " + peerID)
+	n.JoinChannel(testchannelID3, orderer, peer)
+
+	By("waiting for all peers to have the same ledger height")
+	channelPeers := n.PeersWithChannel(channelID)
+	maxLedgerHeight := nwo.GetMaxLedgerHeight(n, channelID, channelPeers...)
+	nwo.WaitUntilEqualLedgerHeight(n, channelID, maxLedgerHeight, channelPeers...)
+
+	helper.invokeMarblesChaincode(chaincode.Name, peer, "transferMarble", testKey, "newowner_update_"+newCCSequence)
+
```

Comment:
It will be cleaner if we keep the helper functions free from hard-coded data and pass the test data from the main flow - similar comment below as well